### PR TITLE
Packages: Reimplement ESLint config as plugin

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to rename and repurpose the current (yet-unpublished) `eslint-config` package to an `eslint-plugin` package. This will allow the package to serve a dual-purpose of providing configurations _and_ custom rules, where the latter is not possible without being published as its own, separate ESLint plugin.

The change in usage is minimal, in regular usage amounting to:

```diff
	extends: [
-		'@wordpress/eslint-config',
+		'plugin:@wordpress/eslint-plugin/recommended',
	],
```

In the course of doing so, I had observed significant divergence, redundancy, and a lack of opportunity for granularity between the bundled rulesets. The proposed changes seek to remedy this by:

- Flattening `configs/rules` into their respective `configs/` files
   - This will also help lend clarity when proper rules are included with the plugins, separate from what are called here as "rulesets" (specific configurations of rules)
- Consolidate the default, ESNext, and ES5 rulesets
   - The default ruleset is now `esnext`, which itself is an extension of `es5`
- Separate out `jsx-a11y` and `react` configurations
   - Enables easy usage of the JavaScript-specific rules without also opting in to an assumed React code environment
   - **Note:** Technically I think the purest approach here would not define `eslint-plugin-react` and `eslint-plugin-jsx-a11y` as dependencies of `eslint-config`, but rather assume them as optional peer dependencies. For the purpose of easing developer "Getting Started" experience, I opted to leave them as dependencies, with the primary downside being that they'd be installed whilst not necessarily used by a consuming project.

**Testing Instructions:**

Verify there is no difference in linting results for the current codebase.

```
npm run lint-js
```

### CODE REVIEW

**Review by ntwb — APPROVED** [review]

**Comment by aduth** [comment]
> Then, over time the `@wordpress/eslint-plugin/recommended` rules can be applied to the legacy codebase, as this will take considerable time IMHO.

Let's plan to consider this as a follow-on I think. This pull request introduces "new" rulesets, but they're really just pulled from what had already existed.

**Comment by gziolo** [comment]
> * **Note:** Technically I think the purest approach here would not define `eslint-plugin-react` and `eslint-plugin-jsx-a11y` as dependencies of `eslint-config`, but rather assume them as optional peer dependencies. For the purpose of easing developer "Getting Started" experience, I opted to leave them as dependencies, with the primary downside being that they'd be installed whilst not necessarily used by a consuming project.

This is a common trade-off in the case when you provide default configuration. I think it's fine as you assume that the default one is going to be used in the majority of cases. I think there are bigger issues if you would like to extend on this. I had the same thought process when proposing `test-e2e` and in particular `puppeteer` dependency. What if someone is not interested in using such script? Fortunately those are `devDependencies` so they don't have any impact on production code.

File: packages/eslint-plugin/configs/esnext.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	],
+	parserOptions: {
+		sourceType: 'module',
+	},
+	rules: {
+		// Disable ES5-specific (extended from ES5)
+		'vars-on-top': 'off',
+
+		// Enable ESNext-specific
+		'array-bracket-spacing': [ 'error', 'always' ],
```
Comment (gziolo) [inline packages/eslint-plugin/configs/esnext.js]: `array-bracket-spacing` is already in ES5 config.

File: packages/eslint-plugin/configs/esnext.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		'arrow-parens': [ 'error', 'always' ],
+		'arrow-spacing': 'error',
+		'computed-property-spacing': [ 'error', 'always' ],
+		'constructor-super': 'error',
+		'no-const-assign': 'error',
+		'no-dupe-class-members': 'error',
+		'no-duplicate-imports': 'error',
+		'no-useless-computed-key': 'error',
+		'no-useless-constructor': 'error',
```
Comment (gziolo) [inline packages/eslint-plugin/configs/esnext.js]: Is `'no-var': 'error',` automatically applied to `es6` env? It's missing when you compare to the previous setup.

**Review by gziolo — APPROVED** [review]
Nice refactoring. Great job! 

There is one rule which is duplicated and can be removed, and one rule I wasn't sure about. Otherwise, it's good to go.

**Comment by aduth** [comment]
Eagle eye, @gziolo 😄 Fixed up both of these. Also ran through once again the prior config to see if there were other missing rules (there weren't, as best I could tell).

**Pull request merged**